File: regular_assignments/day_008/smallest_subarray_with_max_and_min_element/solution.rb
class Solution
  def call
    solve([1, 3, 2]) # 2
    solve([2, 6, 1, 6, 9]) # 3
  end

  def solve(arr)
    min, max = find_min_max(arr)
    return 1 if min == max

    last_min_indx = -1
    last_max_indx = -1
    ans = arr.length
    i = 0
    while i < arr.length
      if arr[i] == max
        if last_min_indx != -1
          ans = find_min(ans, i - last_min_indx + 1)
        end
        last_max_indx = i
      elsif arr[i] == min
        if last_max_indx != -1
          ans = find_min(ans, i - last_max_indx + 1)
        end
        last_min_indx = i
      else
        ans = find_min(ans, arr.length)
      end

      i += 1
    end
    ans
  end

  def find_min_max(arr)
    inf = 10000000000
    min = inf
    max = -inf
    i = 0
    while i < arr.length
      min = find_min(arr[i], min)
      max = find_max(arr[i], max)
      i += 1
    end
    [min, max]
  end

  def find_min(a, b)
    a > b ? b : a
  end

  def find_max(a, b)
    a > b ? a : b
  end
end

# puts '-' * 100
# puts Solution.new.solve([1, 3, 2])
# puts '-' * 100
# puts Solution.new.solve([2, 6, 1, 6, 9])

File: regular_assignments/day_008/smallest_subarray_with_max_and_min_element/solution.go
package smallestSubArrayWithMinMaxElements

func Call() {
	solve([]int{1, 3, 2})
	solve([]int{2, 6, 1, 6, 9})
}

func solve(arr []int) int {
	min, max := findMinMax(arr)
	if min == max {
		return 1
	}
	lastMinIndx := -1
	lastMaxIndx := -1
	ans := len(arr)
	for i := 0; i < ans; i++ {
		if arr[i] == max {
			if lastMinIndx != -1 {
				ans = findMin(ans, i-lastMinIndx+1)
			}
			lastMaxIndx = i
		} else if arr[i] == min {
			if lastMaxIndx != -1 {
				ans = findMin(ans, i-lastMaxIndx+1)
			}
			lastMinIndx = i
		} else {
			ans = findMin(ans, len(arr))
		}
	}
	return ans
}

func findMinMax(arr []int) (int, int) {
	inf := 10000000000
	min := inf
	max := -inf
	for i := 0; i < len(arr); i++ {
		min = findMin(arr[i], min)
		max = findMax(arr[i], max)
	}
	return min, max
}

func findMin(a, b int) int {
	if a > b {
		return b
	} else {
		return a
	}
}

func findMax(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}
